### PR TITLE
Add warp-aggregated atomic increment

### PR DESCRIPTION
**Pull request by ahendriksen**
Implement fast atomic counter increment using warp-aggregated atomics. Useful
for filtering.

Adapted from:
https://developer.nvidia.com/blog/cuda-pro-tip-optimized-filtering-warp-aggregated-atomics/

### CODE REVIEW

**Comment by ahendriksen** [comment]
Moving this functionality from cuml to raft as discussed in https://github.com/rapidsai/cuml/pull/4803

**Review by tfeher — CHANGES_REQUESTED** [review]
Looks good. Please add a unit test.

**Comment by ahendriksen** [comment]
I have added a unit test.

**Review by tfeher — APPROVED** [review]
Thanks @ahendriksen for adding the test! The PR looks good to me.

File: cpp/test/device_atomics.cu [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  rmm::device_scalar<int> counter{0, s};
+  rmm::device_uvector<int> out_device{num_elts, s};
+  std::array<int, num_elts> out_host{0};
+
+  // Write all 1M thread indices to a unique location in `out_device`
+  test_atomic_inc_warp_kernel<<<num_blocks, threads_per_block, 0, s>>>(counter.data(),
+                                                                       out_device.data());
+
+  // Copy data to host
+  RAFT_CUDA_TRY(cudaMemcpy(out_host.data(),
```
Comment (cjnolet) [inline cpp/test/device_atomics.cu:56]: Can we use `raft::copy` here?

**Review by cjnolet — APPROVED** [review]
LGTM. Just a tiny question / nitpick in the test. Thanks @ahendriksen!

**Comment by cjnolet** [comment]
@gpucibot merge

**Pull request merged**